File: 01.RideSharingApp/rideSharing.cpp
#include<bits/stdc++.h>

using namespace std;

class Person {
    public:
    string name;
};

class Driver: private Person {
    Driver(string name) {
        this->name = name;
    }

};

class Rider: private Person {
    private: 
    vector<Ride> completedRides;
    Ride currentRide;

    public:
    Rider(string name) {
        this->name = name;
    }
    void createRide(int, int, int, int);
    void updateRide(int, int, int, int);
    void withdrawRide(int);
    int closeRide();
};

enum RideStatus {IDLE, CREATED, WITHDRAWN, COMPLETED };

class Ride {
    public:
    Ride(){
        id = origin = dest = seats = 0;
        rideStatus = RideStatus::IDLE;
    }
    static const int AMT_PER_KM = 20;   // Amount per kilometer
    int calculateFare(bool);    //boolean is to check if the user is previliged or not

    int getId()const;   // getter for Id
    void setId(int id); // setter for Id

    void setOrigin(int origin); //setter for origin

    void setDest(int dest); // setter for destination

    void setSeats(int seats); // setter for seats

    RideStatus getRideStatus(void)const; // getter for ride status
    void setRideStatus(RideStatus rideStatus); // setter for ride status


    private:
    int id; // ride id
    int origin;
    int dest;
    int seats;
    RideStatus rideStatus;
};
/************* Setter and Getters for Ride class *****************/
int Ride::getId() const {
    return id;
}
void Ride::setId(int id) {
    this->id = id;
}

void Ride::setOrigin(int origin) {
    this->origin = origin;
}

void Ride::setDest(int dest) {
    this->dest = dest;
}

void Ride::setSeats(int seats) {
    this->seats = seats;
}

RideStatus Ride::getRideStatus() const {
    return rideStatus;
}

void Ride::setRideStatus(RideStatus rideStatus) {
    this->rideStatus = rideStatus;
}
/************************************************************/

/*********** Functions from Ride class *********************/
int Ride::calculateFare(bool isPriorityRider) {
    int distance = dest - origin;
    if(seats >= 2) {
        return distance * AMT_PER_KM * seats * (isPriorityRider ? 0.5 : 0.75);
    }
    return distance * AMT_PER_KM * (isPriorityRider ? 0.75 : 1);
}
/***********************************************************/

/************ Functions of rider class *********************/
void Rider::createRide(int id, int origin, int destination, int seats) {
    if(origin >= destination) {
        cout<<"Error! Wrong parameters passes, origin can't be greater than destination!"<<endl;
        return;
    }
    currentRide.setId(id);
    currentRide.setOrigin(origin);
    currentRide.setDest(destination);
    currentRide.setSeats(seats);
    currentRide.setRideStatus(RideStatus::CREATED);
}

void Rider::updateRide(int id, int origin, int destination, int seats) {
    if(currentRide.getRideStatus() == RideStatus::WITHDRAWN) {
        cout<<"Can't update the withdrawn ride!"<<endl;
        return;
    }
    if(currentRide.getRideStatus() == RideStatus::COMPLETED) {
        cout<<"Can't update the completed ride!"<<endl;
        return;
    }
    createRide(id, origin, destination, seats);
}

void Rider::withdrawRide(int id) {
    if(id != currentRide.getId()) {
        cout<<"Unknown rider id"<<endl;
        return;
    }
    if(currentRide.getRideStatus() != RideStatus::CREATED) {
        cout<<"Ride wasn't in progress. Can't withdraw ride"<<endl;
        return;
    }
    currentRide.setRideStatus(RideStatus::WITHDRAWN);
}

int Rider::closeRide() {
    if(currentRide.getRideStatus() != RideStatus::CREATED) {
        cout<<"Can't close the ride which is not in progress!"<<endl;
        return 0;
    }
    currentRide.setRideStatus(RideStatus::COMPLETED);
    completedRides.push_back(currentRide);
    return currentRide.calculateFare(completedRides.size() >= 10);
}
/***********************************************************/

int main() {
    Rider* r = new Rider("kishore");

}

File: 01.RideSharingApp/rideSharing.ts
class Person {
    protected name: string;
    constructor(name: string) {
        this.name = name;
    }
};

class Rider extends Person {
    constructor(name: string) {
        super(name);
    }
    private currentRide: Ride;
    private completedRides: Ride[] = [];

    createRide(id: number, origin: number, destination: number, seats: number) {
        if(origin >= destination) {
            console.error("Error! Destination must be greater than origin!");
            return;
        }
        this.currentRide = new Ride();
        this.currentRide.id = id;
        this.currentRide.origin = origin;
        this.currentRide.destination = destination;
        this.currentRide.rideStatus = RideStatus.CREATED;
    }

    updateRide(id: number, origin: number, destination: number, seats: number) {
        if(this.currentRide.rideStatus == RideStatus.WITHDRAWN) {
            console.error("Can't update the ride which is withdrawn");
            return;
        }
        if(this.currentRide.rideStatus == RideStatus.COMPLETE) {
            console.error("Can't update the ride which is complete");
            return;
        }

        this.createRide(id, origin, destination, seats);
    }

    withdrawRide(id: number) {
        if(id != this.currentRide.id) {
            console.log("Unknown ID");
            return;
        }
        if(this.currentRide.rideStatus != RideStatus.CREATED) {
            console.log("Can only withdraw a ride which is in progress!!");
            return;
        }
        this.currentRide.rideStatus = RideStatus.WITHDRAWN;
    }

    closeRide(): number {
        if(this.currentRide.rideStatus != RideStatus.CREATED) {
            console.log("Can only close a ride which is in progress!!");
            return 0;
        }
        this.currentRide.rideStatus = RideStatus.COMPLETE;
        this.completedRides.push(this.currentRide);
        return this.currentRide.calculateFare(this.completedRides.length >= 10);
    } 
};

class Driver extends Person {
    constructor(name: string) {
        super(name);
    }
};

enum RideStatus {CREATED, IDLE, WITHDRAWN, COMPLETE};

class Ride {
    static amountPerKM = 20;
    private _id: number;
    private _origin: number;
    private _destination: number;
    private _seats: number;
    private _rideStatus: RideStatus;
    constructor(){
        this._id = this._origin = this._destination = this._seats = 0;
        this._rideStatus = RideStatus.IDLE;
    }
    get id():number {
        return this._id;
    }
    set id(rideId: number) {
        this._id = rideId;
    }
    set origin(rideOrigin: number) {
        this._origin = rideOrigin;
    }
    set destination(rideDestination: number) {
        this._destination = rideDestination;
    }
    set seats(rideSeats: number) {
        this._seats = rideSeats;
    }
    get rideStatus():RideStatus {
        return this._rideStatus;
    }
    set rideStatus(status: RideStatus) {
        this._rideStatus = status;
    }
    calculateFare(isPrivileged: boolean): number {
        let distance = this._destination - this._origin; 
        if(this._seats >= 2) {
            return this._seats * distance * Ride.amountPerKM * (isPrivileged ? 0.5 : 0.75);
        }
        return distance * Ride.amountPerKM * (isPrivileged ? 0.75 : 1);
    }
};

function main() {
    let rider = new Rider('Lucifer');
    let driver = new Driver('Amenadeil');

}

main();